### PR TITLE
Proxy connector

### PR DESCRIPTION
**Pull request by massand**
- Added proxy client support
- Refactored Hub client to reduce redundancy
- Verified manually using SaS and X.509 auth with squid proxy 

Note that the hyper-proxy crate is still pinned to https://github.com/massand/hyper-proxy in this commit. However, the final crate source will be https://github.com/Azure/hyper-proxy after the fork is checked into main there.

### CODE REVIEW

File: http-common/src/proxy.rs
Comment (arsing) [inline http-common/src/proxy.rs:139]: `C: hyper::client::connect::Connect` should let you skip most / all of these bounds.

**Review by arsing — CHANGES_REQUESTED** [review]

File: http-common/src/proxy.rs
Comment (daprilik) [inline http-common/src/proxy.rs:13]: You already have a `#[cfg(feature = "tokio1")]` block on the `mod proxy` declaration in `http-common/src/lib.rs`, so there's no reason to clutter up this file with a whole bunch of redundant `#[cfg(feature = "tokio1")]` declarations.

File: http-common/src/proxy.rs
Comment (daprilik) [inline http-common/src/proxy.rs:11]: nit: import ordering + formatting.

- split `std` into individual lines (i.e: separate `use std::future::Future`, `use std::io`, etc...)
- standard library imports before external libraries

File: http-common/src/proxy.rs
Comment (daprilik) [inline http-common/src/proxy.rs:149]: nit: you have `use std::task::Poll` in scope, so no need to use the fully qualified path here (same with `Context`)

File: http-common/src/proxy.rs
Comment (massand) [reply http-common/src/proxy.rs:159]: I was running into issues with the inner connector lifetime since it isn't cloneable.

Comment (massand) [reply http-common/src/proxy.rs:159]: Never mind - it actually is cloneable. I'll fix it

File: http-common/src/proxy.rs
Comment (massand) [reply http-common/src/proxy.rs:139]: So I tried this, but I am missing something here: 
```
impl<C> hyper::service::Service<http::uri::Uri> for MaybeProxyConnector<C>
where
    C: hyper::client::connect::Connect + Clone + 'static,
{
    type Response = MaybeProxyStream<C::Response>;
    type Error = io::Error;
    type Future = Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>> + Send>>;
```
However, this impacts the `Response`, since `Connect` doesn't have a `Response`. Are you suggesting I remove the `Service` implementation and implement `Connect` instead?

Comment (arsing) [reply http-common/src/proxy.rs:139]: >However, this impacts the `Response`, since `Connect` doesn't have a `Response`.

Ah right, `Connect` for `Service` is a blanket impl instead of inheritance, so trait bounds won't inherit the way I was thinking of anyway.

File: http-common/Cargo.toml
Comment (massand) [reply http-common/Cargo.toml]: Done.

File: http-common/Cargo.toml
Comment (massand) [reply http-common/Cargo.toml]: Done.

File: http-common/src/proxy.rs
Comment (massand) [reply http-common/src/proxy.rs]: Done.

File: identity/aziot-cloud-client-async-common/Cargo.toml
Comment (massand) [reply identity/aziot-cloud-client-async-common/Cargo.toml]: Done.

File: identity/aziot-identityd/src/lib.rs
Comment (massand) [reply identity/aziot-identityd/src/lib.rs]: I want to do a separate PR for it since I need to change the API to pass in trust cert bundle for EST connection

File: http-common/src/proxy.rs
Comment (massand) [reply http-common/src/proxy.rs]: Done.

File: http-common/src/proxy.rs
Comment (massand) [reply http-common/src/proxy.rs]: Done.

File: http-common/src/proxy.rs
Comment (massand) [reply http-common/src/proxy.rs]: Done.

File: http-common/Cargo.toml
Comment (daprilik) [inline http-common/Cargo.toml:14]: Will there be a follow-up PR that switches this over to https://github.com/Azure/hyper-proxy, or is this PR blocked until the fork PR is approved?

Comment (massand) [reply http-common/Cargo.toml:14]: We can wait for the fork update. We rather use the final repo in our dependencies. If you have other comments, let me know.

Comment (massand) [reply http-common/Cargo.toml]: Updated.

File: http-common/src/proxy.rs
Comment (arsing) [inline http-common/src/proxy.rs:60]: Add `poll_write_vectored` and `is_vectored`

File: http-common/Cargo.toml
Comment (arsing) [inline http-common/Cargo.toml:24]: You don't seem to need this.

File: http-common/src/proxy.rs
Comment (arsing) [inline http-common/src/proxy.rs:170]: You don't need to clone. Also you can use shorter names for the bindings like `c` or `connector`.

```rust
MaybeProxyConnector::NoProxy(c) => {
    let stream = https_connector.call(req);
    Box::pin(async {
        let stream = stream.await.map_err(...)?;
        Ok(...)
    })
}
```

File: http-common/src/proxy.rs
Comment (arsing) [inline http-common/src/proxy.rs:163]: `e` is already a `std::io::Error`, so you don't need this entire `match`. Just let the first one return its value without binding it.

**Review by arsing — CHANGES_REQUESTED** [review]

File: http-common/src/proxy.rs
Comment (massand) [reply http-common/src/proxy.rs]: Fixed.

File: http-common/src/proxy.rs
Comment (massand) [reply http-common/src/proxy.rs]: Fixed.

File: http-common/Cargo.toml
Comment (massand) [reply http-common/Cargo.toml]: Fixed.

File: http-common/src/proxy.rs
Comment (massand) [reply http-common/src/proxy.rs:78]: Fixed.

**Review by arsing — APPROVED** [review]

**Pull request merged**
